### DISCUSSION TITLE
Animating prims in the USD stage

### DISCUSSION THREAD

**Discussion by RichardFrangenberg** [Autodesk/maya-usd#1320] [Q&A]
I'm referencing some assets into my USD stage and want to add a simple animation to them. But it looks like I can't add keyframes to the xform prims.
Is there any way to keyframe USD assets in Maya?

**Comment by kxl-adsk** [comment]
Hi Richard.

The animation wasn't the goal for Maya 2022 release, but having said that we have investigated and proposed a foundation that can enable workflows like:
* parenting (e.g. DAG to Prim)
* animation (i.e. driving prim attributes with Maya animation curves)
* attribute level constraints

We don't have nice workflows for them exposed yet, but there is a way with custom tooling to use it. Since now Maya 2022 is released, I will try to provide a better introduction to this work and post it here. In the meanwhile if you are willing to read some python code you could explore different workflows regression test is validation - https://github.com/Autodesk/maya-usd/blob/dev/test/lib/testMayaUsdProxyAccessor.py#L39 (there is one for keyframing)

Cheers,
KxL

**Reply by RichardFrangenberg** [reply]
Thanks for the answer. Great to hear that it's possible with custom tools. Looking forward to the official implementation though.
Cheers,
Richard

**Comment by brycegbrazen** [comment]
@womanyoyoyo-adsk bringing this thread back up, is the in-USD animation workflow on the roadmap for you guys right now? Just was thinking about the possibility of having a USD Camera in Maya (with some basic parenting under NurbsCurves for offsets), and the viability of a workflow where a camera is kept in USD at all times between departments.
This would include being able to key/animate transforms as well as other camera attributes that are present in the Attribute Editor already such as focal length and clipping range. Ideally for the transforms we could see the keys in the timeline as well.

**Reply by wallworm** [reply]
We definitely hold animation workflows at the top of our priorities and we are continuing to work in this area. Can you share your ideal workflows more in-depth? Would you be willing to walk our designers through the scenarios that are important to you?

**Reply by richardHaw** [reply]
any updates on this?

**Reply by wallworm** [reply]
There is no specific update at this time but our current plan is to implement [Pixar's USD Anim](https://github.com/Pixneb/OpenUSD-proposals/blob/spline-animation-review/proposals/spline-animation/README.md) when it is implemented in USD. See the current proposal [here](https://github.com/Pixneb/OpenUSD-proposals/blob/spline-animation-review/proposals/spline-animation/README.md).
